Reset `_closeables` in `_init` for flow reuse

alxtkr77 opened the pull request.

Commit:
Reset _closeables in _init for flow reuse

gtopper reviewed (commented):
Looks great. Would be good to add a regression unit test.

Commit:
Add unit tests for _closeables reset in _init for flow reuse

Tests verify that _closeables is properly reset when flows are reused:
- test_flow_reuse_resets_closeables: Base Flow class resets to []
- test_aggregate_by_key_reuse_resets_closeables: AggregateByKey sets [table]
- test_map_with_state_reuse_resets_closeables: MapWithState sets [state] for Table
- test_map_with_state_no_closeables_without_close_method: No closeables for dict state
- test_nosql_target_reuse_resets_closeables: NoSqlTarget sets [table]

Commit:
Fix unit tests to properly detect ML-11518 bug and add clarifying comments

Updated tests to check source._closeables instead of downstream step's _closeables,
as the bug manifests in upstream steps. Added comments clarifying which tests detect
the ML-11518 regression (3 tests) vs baseline/edge case tests (2 tests).

Verified:
- 3 tests FAIL on commit 74fec1a (before fix) - correctly detecting bug
- All 5 tests PASS on commit 321dfec (with fix) and current branch

gtopper commented on tests/test_flow.py:
```diff
@@ -5636,3 +5636,191 @@ def test_flow_reuse_with_cycle():
+
+    Baseline test for flow reuse without closeables. This test uses steps (Map, Reduce)
+    that don't create closeables, so it verifies the baseline behavior. Note: This test
+    does NOT detect the ML-11518 bug since there are no closeables to accumulate.
```
That's a nice note for the PR, but not really a relevant comment once merged.

gtopper commented on tests/test_flow.py:
```diff
@@ -5636,3 +5636,191 @@ def test_flow_reuse_with_cycle():
+        for i in range(5):
+            controller.emit(i)
+        controller.terminate()
+        result = controller.await_termination()
```
These need to run on assertion failure (or even an unexpected error), or the test will hang. This applies to all the tests in this PR. Termination needs to be done in a finally block.

gtopper commented on tests/test_flow.py:
```diff
@@ -5636,3 +5636,191 @@ def test_flow_reuse_with_cycle():
+            assert len(source._closeables) == first_run_closeables_count, (
+                f"Run {run_num}: source._closeables count should be {first_run_closeables_count}, "
+                f"got {len(source._closeables)} (closeables are accumulating!)"
+            )
```
This is a very long-winded way of just doing
```suggestion
    for run_num in range(3):
        controller = source.run()
        assert len(source._closeables) == 1
```

gtopper commented on tests/test_flow.py:
```diff
@@ -5636,3 +5636,191 @@ def test_flow_reuse_with_cycle():
+    This test DOES detect the bug by checking source._closeables remains consistent.
+    """
+    from storey import AggregateByKey, FieldAggregator
+    from storey.dtypes import FixedWindows
```
I guess tests for aggregates should actually go in `test_aggregate_by_key.py`. And imports done at the top.

Commit:
Simplify ML-11518 regression tests based on code review

- Replace dynamic closeables tracking with direct assertions
- Add try/finally blocks to prevent test hangs on failure
- Remove PR-specific docstring notes that don't apply post-merge
- Move AggregateByKey test to test_aggregate_by_key.py

gtopper reviewed (approved):
Cool. Please rebase / merge from dev before merge. This branch is missing
```
commit 0bfaa08e78570c2909275dae2740829a056f3aa8 (mlrun/development, mlrun/HEAD, development)
Author: davesh0812 <85231462+davesh0812@users.noreply.github.com>
Date:   Sun Dec 21 12:33:03 2025 +0200

    Minor fixes to the cyclic pr (#598)
```
